src/util/clustering.js

```diff
@@ -0,0 +1,265 @@
+/**
+ * @file
+ * Using methods adapted from leaflet to cluster an array of positions
+ * hierarchically given an array of length scales (zoom levels).
+ */
+
+(function () {
+  'use strict';
+
+  /**
+   * This class manages a group of nearby points that are clustered as a
+   * single object for display purposes.  The class constructor is private
+   * and only meant to be created by the ClusterGroup object.
+   *
+   * This is a tree-like data structure.  Each node in the tree is a
+   * cluster containing child clusters and unclustered points.
+   *
+   * @class
+   * @private
+   *
+   * @param {geo.util.ClusterGroup} group The source cluster group
+   * @param {number} zoom The zoom level of the current node
+   * @param {object[]} children An array of ClusterTrees or point objects
+   */
+  function ClusterTree(group, zoom, children) {
+    this._group = group;
+    this._zoom = zoom;
+    this._points = [];     // Unclustered points
+    this._clusters = [];   // Child clusters
+    this._count = 0;       // Total number of points
+    this._parent = null;
+    this._coord = null;    // The cached coordinates
+    var that = this;
+
+    // add the children provided in the constructor call
+    (children || []).forEach(function (c) {
+      that._add(c);
+    });
+  }
+
+  /**
+   * Add a point or cluster as a child to the current cluster.
+   * @param {object} pt A ClusterTree or point object
+   * @private
+   */
+  ClusterTree.prototype._add = function (pt) {
```

Comment:
Yeah, I was reading about the ES6 Class keyword. Looks interesting but I haven't looked into how it solves some of the most complicated inheritance in JS. I think it would be nice to think about refactoring our class syntax when it comes out though. For few reasons, I am not a big fan of prototype methods and Class seems to be a better option.